### PR TITLE
(closes #1508) allow PSyIR backend checks to be disabled

### LINKED ISSUE #1508: Add option to disable global checks in backend
At the moment transformations such as `ACCKernelsTrans` allow the user to disable the check for valid node types when calling `apply()`. However, there are further checks performed in the backend when code is generated which means that it is still not possible to generate code for such a case (without manually doing it in e.g. the transformation script). For consistency (and ease of use) we therefore need a way of disabling the checks in the backend. Two options are:

 1. A configuration-file option;
 2. A command-line flag.

Given that the checks are there for a reason, I think we would only want to disable them for particular files (or even routines). I therefore favour the command-line flag option.

### PR DESCRIPTION
**Pull request by arporter**
As the title says. It also includes a small fix to the validation of `ACCLoopDirective` to permit it to appear on its own in a Routine so long as that Routine also has an `ACCRoutineDirective`.

### CODE REVIEW

**Comment by arporter** [comment]
I need to add a test for the tweak to the ACCDirective validation in the PSyIR backend - that they are permitted in a Routine with an `ACCRoutineDirective`.

**Comment by arporter** [comment]
This is ready for review now. It provides an (optional and so backwards-compatible) command-line flag and corresponding configuration-file entry to control whether or not the backend does validation. I've updated the associated documentation.
It also fixes a small issue I found with that validation when adding an ACC loop directive to a subroutine.

One for @rupertford, @hiker or @TeranIvy.

**Comment by sergisiso** [comment]
> As the title says. It also includes a small fix to the validation of ACCLoopDirective to permit it to appear on its own in a Routine so long as that Routine also has an ACCRoutineDirective.

I have been using this, but ACCRoutineDirective adds "acc routine" without clauses that defaults to seq, so the "acc loop independent doesn't do anything (I believe). We need to implement ACCRoutineDirective propterty to set the clause to "acc routine gang/worker/vector" and then the compiler will take advantage of the ACCLoop. Is this right? Is this coming in another PR?

**Comment by arporter** [comment]
> I have been using this, but ACCRoutineDirective adds "acc routine" without clauses that defaults to seq, so the "acc loop independent" doesn't do anything (I believe).

I'm not sure although I think it just means that the body of the routine will be executed in serial. That doesn't prevent each separate call to the routine being executed in parallel though (due to the `ACC loop` on the outer loop over cells) and I think I see that in the profiler (for Gravity Wave) - it would be very, very slow if the whole thing was serial. If you mean, applying `ACC loop` to the `k` loop within the kernel then that's different - I haven't tried to do that yet (I got bogged down in getting `atomic` to work reliably).

**Comment by sergisiso** [comment]
> I think it just means that the body of the routine will be executed in serial. That doesn't prevent each separate call to the routine being executed in parallel though

Yes, I agree. I just thought that you were going for parallelism inside the loop for your comment about "a small fix to the validation of ACCLoopDirective to permit it to appear on its own in a Routine" but I agree cells and atomics is a good initial target.

**Comment by arporter** [comment]
Taking this back as I've realised that I need to update `f2pygen.PSyIRGen` too.

**Comment by arporter** [comment]
OK, ready for review now.

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 1]
```diff
@@ -518,18 +518,26 @@ def validate_global_constraints(self):
...
-                            within some OpenACC parallel or kernels region.
+        :raises GenerationError: if this ACCLoopDirective is not enclosed
+            within some OpenACC parallel or kernels region and is not in a
+            Routine that has been marked up with an 'ACC Routine' directive.
         '''
         # It is only at the point of code generation that we can check for
         # correctness (given that we don't mandate the order that a user can
         # apply transformations to the code). As an orphaned loop directive,
         # we must have an ACCParallelDirective or an ACCKernelsDirective as
         # an ancestor somewhere back up the tree.
```
Comment (sergisiso) [inline src/psyclone/psyir/nodes/acc_directives.py:529]: The ACCRoutine case is missing from this comment, but tbh I would just remove all this comment as this is already explained in the docstring and the code is quite clear.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter The changes look good, I just made some minor comments.

I think I understood this from the documentation, but just to double check it with you, the new option is just to simplify the developers life, ultimately the atomics validation issue that you mentioned in the issue will be handled properly in the global constraints and we won't need to disable the backends checks, right?

File: doc/user_guide/configuration.rst [hunk 2]
```diff
@@ -162,26 +164,30 @@ DEFAULTAPI              The API that PSyclone assumes an Algorithm/Kernel
...
                         "psyir_tmp".
-VALID_PSY_DATA_PREFIXES Which class prefixes are permitted in any
+VALID_PSY_DATA_PREFIXES Which class prefixes are permitted in any               list of str
                         PSyData-related transformations. See :ref:`psy_data`
                         for details.
-======================= =======================================================
+BACKEND_CHECKS_ENABLED  Optional (defaults to True). Whether or not the PSyIR   bool
+                        backend should validate the tree that it is passed.
+			Can be overridden by the ``--backend`` command-line
+			flag (see :ref:`backend-options`).
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/configuration.rst:189]: Done.

File: doc/developer_guide/psyir_backends.rst [hunk 3]
```diff
@@ -34,6 +34,7 @@
    Authors: R. W. Ford, A. R. Porter, S. Siso and N. Nobre, STFC Daresbury Lab
 
 
+.. _psyir-backends:
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developer_guide/psyir_backends.rst:37]: Done.

File: doc/user_guide/psyclone_command.rst [hunk 4]
```diff
@@ -379,6 +385,30 @@ by the ``-I``/``--include`` flags. (Currently this search assumes that a
 module named e.g. "my_mod" will be in a file named "my_mod.*90" - see issue
 #1895.)
 
+.. _backend-options:
+
+Backend Options
+---------------
+
+The final code generated by PSyclone is created by passing a PSyIR
+tree to one of the 'backends' (see :ref:`dev_guide:psyir-backends` in
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/psyclone_command.rst:394]: No, this is handled auto-magically by Sphinx. It won't work locally until the associated change to the Dev Guide is on RTD but the CI tests are cleverer and check the links against the version of the Dev Guide in the PR itself. If they are happy then all is well.

File: src/psyclone/psyir/nodes/acc_directives.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_directives.py:529]: 'tis gone.

File: doc/developer_guide/psyir_backends.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developer_guide/psyir_backends.rst:37]: (and in the other docs files that needed it)

**Comment by arporter** [comment]
Ready for another look assuming CI happiness.

**Review by sergisiso — APPROVED** [review]
@arporter All the comments have been addressed (but the tabs->spaces but I will do it myself). Tests, codecov, documentation all are green. I am running integration tests just in case while the other PRs pending to merge are being handled.

**Pull request merged**